### ISSUE TITLE
Support for resolving Yarn Plug 'n Play modules

### ISSUE THREAD

**Issue by empyrical** [react/metro#308]
**Do you want to request a *feature* or report a *bug*?**
Request a feature

**What is the current behavior?**
Metro does not (yet) have PnP support

**If the current behavior is a bug, please provide the steps to reproduce and a minimal repository on GitHub that we can `yarn install` and `yarn test`.**

**What is the expected behavior?**
I think it would be very handy if Metro had support for resolving Yarn PnP modules.

Not only would there be the benefits you'd get for using it in other project, I believe that if [this Yarn feature request](https://github.com/yarnpkg/yarn/issues/6642) is able to be resolved, this could be a reasonable alternative to using symlinks (since they are [not able to be supported](https://github.com/facebook/metro/issues/1) for technical reasons)

**Comment by arcanis** [comment]
It actually already does! We just need to open-source the resolver in a separate package (or make it native, possible) 😃 ping @rafeca

**Comment by vjpr** [comment]
I found it implemented here: https://github.com/facebook/metro/commit/60bad942f6f32d2ac1af7dc58e3e768a9aee1752

**Comment by dalinarkholin** [comment]
@arcanis @empyrical was Pnp support removed? Im guessing that is why this is still open?

**Comment by arcanis** [comment]
Yep, I think FB removed it about a year ago since they don't use it internally. You should be able to add it back by overriding `resolveRequest` in your configuration, though.

**Comment by arlyon** [comment]
Now that yarn uses pnp by default, can we consider adding it again? (esp given that yarn is (was) a facebook project)

**Comment by dalinarkholin** [comment]
@arlyon I just added an example repo with pnp and react native working. Checkout the metro.config.js here: [https://github.com/dalinarkholin/react-native-pnp-enabled](https://github.com/dalinarkholin/react-native-pnp-enabled)

Also my comment here: [https://github.com/react-native-community/cli/issues/27#issuecomment-625912523](https://github.com/react-native-community/cli/issues/27#issuecomment-625912523)

**Comment by Dirklectisch** [comment]
Would it help if I provide a pull request to get this back in? If there is no interest in adding this back in we could create a separate package for it so that at least React Native has a way forward for adopting Yarn 2.

**Comment by dvicory** [comment]
> Would it help if I provide a pull request to get this back in? If there is no interest in adding this back in we could create a separate package for it so that at least React Native has a way forward for adopting Yarn 2.

I'd love to see this either way, but would use a separate package if metro isn't interested in adding it.

**Comment by hbiede** [comment]
Is there anything a contributor could do to move this along (question aimed at repo maintainers)? I’d love to see this as a feature supported by default since my company is wanting to make the move to yarn 2 for web projects

**Comment by evelant** [comment]
Since it's 3 years and counting with no movement on this I"m guessing it's never going to happen. The expo crew seems to worked around it. See here: https://github.com/expo/expo/tree/master/packages/expo-yarn-workspaces

**Comment by alamothe** [comment]
@evelant Yarn PNP or no?

**Comment by evelant** [comment]
@alamothe yes, the latest yarn with default settings. You'll just need to follow the docs there and add some package names to the symlinks field for poorly configured packages that hardcode paths.

**Comment by Slapbox** [comment]
Any update? Yarn 2 has been stable and widely used for years. It would be nice if the React Native ecosystem could get on-board.

**Comment by robhogan** [comment]
Metro now (since 0.79.0's default-on symlink support) fully supports modern package managers, including Yarn 2+ with `nodeLinker: pnpm` and `nodeLinker: node-modules`. I believe that's most of the ask from folks in this thread. In particular, Metro fully supports workspaces now.

The specific PnP feature (`nodeLinker: pnp`) is *not* supported (it was briefly, between 60bad942f, v0.50.0, and 1ff21a8ff, v0.55.0). It doesn't produce Node.js-compatible node_modules layouts and has its own algorithm that always needs bespoke support. Some other bundlers support it, some (eg TurboPack) don't.

Supporting it *correctly* according to Metro's architecture isn't trivial, and isn't a priority.

Closing as not planned. If folks still want PnP specifically even now that we support symlinks, workspaces etc, please open a new issue. If there's still sufficient interest we might reconsider.

**Comment by Nantris** [comment]
Thank you for the update @robhogan.

I created #1928 for people still interested in PnP support (as I certainly am).
